From 8ed70b46247fa14cbed6dc6702b8227abf22b119 Mon Sep 17 00:00:00 2001
From: Marcin Niestroj <m.niestroj@emb.dev>
Date: Mon, 6 May 2024 08:36:38 +0200
Subject: [PATCH] drivers: nsos: support ioctl(FIONREAD)

This further increases compatibility with tests defined in
'tests/net/socket/udp/'.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>
---
 drivers/net/nsos.h         |  2 ++
 drivers/net/nsos_adapt.c   | 13 +++++++++++++
 drivers/net/nsos_sockets.c |  9 +++++++++
 3 files changed, 24 insertions(+)

diff --git a/drivers/net/nsos.h b/drivers/net/nsos.h
index bafb40542e0..911e49dff2c 100644
--- a/drivers/net/nsos.h
+++ b/drivers/net/nsos.h
@@ -138,6 +138,8 @@ void nsos_adapt_poll_update(struct nsos_mid_pollfd *pollfd);
 int nsos_adapt_fcntl_getfl(int fd);
 int nsos_adapt_fcntl_setfl(int fd, int flags);
 
+int nsos_adapt_fionread(int fd, int *avail);
+
 int nsos_adapt_getaddrinfo(const char *node, const char *service,
 			   const struct nsos_mid_addrinfo *hints,
 			   struct nsos_mid_addrinfo **res,
diff --git a/drivers/net/nsos_adapt.c b/drivers/net/nsos_adapt.c
index d3f787762c4..7e3363c73f8 100644
--- a/drivers/net/nsos_adapt.c
+++ b/drivers/net/nsos_adapt.c
@@ -20,6 +20,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <sys/epoll.h>
+#include <sys/ioctl.h>
 #include <sys/socket.h>
 #include <unistd.h>
 
@@ -939,6 +940,18 @@ int nsos_adapt_fcntl_setfl(int fd, int flags)
 	return 0;
 }
 
+int nsos_adapt_fionread(int fd, int *avail)
+{
+	int ret;
+
+	ret = ioctl(fd, FIONREAD, avail);
+	if (ret < 0) {
+		return -errno_to_nsos_mid(errno);
+	}
+
+	return 0;
+}
+
 static void nsos_adapt_init(void)
 {
 	nsos_epoll_fd = epoll_create(1);
diff --git a/drivers/net/nsos_sockets.c b/drivers/net/nsos_sockets.c
index 5dd528e9101..df6f97e8745 100644
--- a/drivers/net/nsos_sockets.c
+++ b/drivers/net/nsos_sockets.c
@@ -371,6 +371,15 @@ static int nsos_ioctl(void *obj, unsigned int request, va_list args)
 
 		return -errno_from_nsos_mid(-ret);
 	}
+
+	case ZFD_IOCTL_FIONREAD: {
+		int *avail = va_arg(args, int *);
+		int ret;
+
+		ret = nsos_adapt_fionread(sock->pollfd.fd, avail);
+
+		return -errno_from_nsos_mid(-ret);
+	}
 	}
 
 	return -EINVAL;
-- 
2.45.2

